shared/js/ui/views/whitelist.es6.js

```diff
@@ -15,19 +15,84 @@ function Whitelist (ops) {
 Whitelist.prototype = window.$.extend({},
   Parent.prototype,
   {
+    _isHiddenClass: 'is-hidden',
+    _isDisabledClass: 'is-disabled',
+    _isInvalidInputClass: 'is-invalid-input',
 
     _removeItem: function (e) {
-      var itemIndex = window.$(e.target).data('item')
+      const itemIndex = window.$(e.target).data('item')
+      const $listItem = window.$(e.target).parent()
       this.model.removeDomain(itemIndex)
-      this.setWhitelistFromSettings()
+
+      // No need to rerender the whole view
+      $listItem.remove()
+    },
+
+    _addItem: function (e) {
+      const url = this.$url.val()
+      let isValidInput = false;
+      if (url) {
+        isValidInput = this.model.addDomain(url)
+      }
+
+      if (isValidInput) {
+        this.setWhitelistFromSettings()
+      } else {
+        this._showErrorMessage()
+      }
+    },
+
+    _showErrorMessage: function () {
+        this.$add.addClass(this._isHiddenClass)
+        this.$error.removeClass(this._isHiddenClass)
+        this.$url.addClass(this._isInvalidInputClass)
+    },
+
+    _hideErrorMessage: function () {
+        this.$add.removeClass(this._isHiddenClass)
+        this.$error.addClass(this._isHiddenClass)
+        this.$url.removeClass(this._isInvalidInputClass)
+    },
+
+    _manageInputChange: function (e) {
+      const isButtonDisabled = this.$add.hasClass(this._isDisabledClass)
+
+      this._hideErrorMessage()
+      
+      if (this.$url.val() && isButtonDisabled) {
+        this.$add.removeClass(this._isDisabledClass)
+      } else if (!this.$url.val()) {
+        this.$add.addClass(this._isDisabledClass)
+      }
+
+      if (!isButtonDisabled && e.key === 'Enter') {
+        // also add to whitelist on enter
+        this._addItem()
+      }
+    },
+
+    _showAddToWhitelistInput: function (e) {
+      this.$url.removeClass(this._isHiddenClass)
+      this.$url.focus()
+      this.$add.removeClass(this._isHiddenClass)
+      this.$showadd.addClass(this._isHiddenClass)
+      e.preventDefault()
     },
 
     setup: function () {
-      this._cacheElems('.js-whitelist', ['remove'])
+      this._cacheElems('.js-whitelist', [
+        'remove',
+        'add',
+        'error',
+        'show-add',
+        'url'
+      ])
 
       this.bindEvents([
         [this.$remove, 'click', this._removeItem],
-        [this.store.subscribe, 'action:backgroundMessage', this.update]
```

Comment:
We don't really need this here - plus this would override my change a couple lines above and rerender the whole view when removing a whitelisted domain.